pkg/remotewrite/remotewrite.go

```diff
@@ -49,116 +53,247 @@ func New(params output.Params) (*Output, error) {
 
-	o.logger.WithField("nts", nts).Debug("Converted samples to time series in preparation for sending.")
+	// TODO: maybe some cache?
+	var promSeries []prompb.TimeSeries
+	for _, s := range series {
+		promSeries = append(promSeries, o.formatSeriesAsProm(s)...)
+	}
 
 	req := prompb.WriteRequest{
-		Timeseries: promTimeSeries,
+		Timeseries: promSeries,
+	}
+
+	buf, err := proto.Marshal(&req)
+	if err != nil {
+		o.logger.WithError(err).Fatal("Failed to marshal timeseries")
+		return
+	}
+
+	// TODO: this call can panic, find the source and fix it
+	encoded := snappy.Encode(nil, buf)
+	if err := o.client.Store(context.Background(), encoded); err != nil {
+		o.logger.WithError(err).Error("Failed to store timeseries")
+		return
 	}
+}
+
+// TODO: cache the prom time series (name + tags)
 
-	if buf, err := proto.Marshal(&req); err != nil {
-		o.logger.WithError(err).Fatal("Failed to marshal timeseries.")
-	} else {
-		encoded := snappy.Encode(nil, buf) // this call can panic
-		if err = o.client.Store(context.Background(), encoded); err != nil {
-			o.logger.WithError(err).Error("Failed to store timeseries.")
+func (o *Output) formatSeriesAsProm(series *tsdb.TimeSeries) []prompb.TimeSeries {
+	labels := tagsToLabels(o.config, series.Tags)
+
+	switch styp := series.Sink.(type) {
+	case *tsdb.CountSeries, *tsdb.GaugeSeries, *tsdb.RateSeries:
+		return []prompb.TimeSeries{
+			{
+				Labels: appendNameLabel(labels, series.MetricName),
+				Samples: []prompb.Sample{
+					{
+						Value:     series.Sink.Value(),
+						Timestamp: timestamp.FromTime(time.Now()),
+					},
+				},
+			},
 		}
+
+	case *tsdb.TrendSeries:
+		return MapTrend(series.MetricName, styp, labels)
+	default:
+		panic("the time series type is not supported")
 	}
 }
 
-func (o *Output) convertToTimeSeries(samplesContainers []metrics.SampleContainer) []prompb.TimeSeries {
-	promTimeSeries := make([]prompb.TimeSeries, 0)
+func (o *Output) sinkSeriesFromSamples(samplesContainers []metrics.SampleContainer) (uint, error) {
+	sinked := make(map[uint64]*tsdb.TimeSeries)
 
+	// TODO: aggregate same time series across containers (maybe caching?)
+	// useless to fetch and update for each single sample
 	for _, samplesContainer := range samplesContainers {
 		samples := samplesContainer.GetSamples()
 
 		for _, sample := range samples {
-			// Prometheus remote write treats each label array in TimeSeries as the same
-			// for all Samples in those TimeSeries (https://github.com/prometheus/prometheus/blob/03d084f8629477907cab39fc3d314b375eeac010/storage/remote/write_handler.go#L75).
-			// But K6 metrics can have different tags per each Sample so in order not to
-			// lose info in tags or assign tags wrongly, let's store each Sample in a different TimeSeries, for now.
-			// This approach also allows to avoid hard to replicate issues with duplicate timestamps.
-
-			labels, err := tagsToLabels(sample.Tags, o.config)
+			tset, err := o.tagSetFromTags(sample.Tags.CloneTags())
```

Comment:
FWIW, this is a bit inefficient, since it will make a copy of the `map[string]string`. Ideally, all of these things should be done with the minimum amount of copying and allocation required. That's only possible on the k6 side, by refactoring the types there directly, so just making a note.